common/scpi.c

```diff
@@ -42,22 +43,38 @@ static struct device *scpi_msgbox;
 /** Buffers used to hold messages while they are being processed. */
 static struct scpi_buffer scpi_buffers[SCPI_BUFFER_MAX];
 
+/**
+ * Allocate a temporary buffer for processing an SCPI message.
+ *
+ * Because this function is called from both IRQ handlers and process context,
+ * it must be locked.
+ */
 static struct scpi_buffer *
 scpi_alloc_buffer(uint8_t client)
 {
+	struct scpi_buffer *buffer = NULL;
+	uint32_t flags;
+
+	flags = disable_interrupts();
 	for (size_t i = 0; i < SCPI_BUFFER_MAX; ++i) {
 		if (!scpi_buffers[i].busy) {
-			struct scpi_buffer *buffer = &scpi_buffers[i];
-
-			/* Mark the buffer as being in use. */
+			buffer         = &scpi_buffers[i];
 			buffer->busy   = true;
 			buffer->client = client;
-
-			return buffer;
+			break;
 		}
 	}
+	restore_interrupts(flags);
 
-	return NULL;
+	return buffer;
+}
+
+static inline void
+scpi_free_buffer(struct scpi_buffer *buffer)
+{
+	/* Prevent moving earlier operations across the unlock operation. */
+	barrier();
```

Comment:
Is there any way you could provide more clarity on what this function does by naming its something else?

Reply:
Not really. That's what this function is: a full compiler barrier. (Incidentally, since we don't have a data cache, it's also a full memory barrier.) See https://en.wikipedia.org/wiki/Memory_barrier

The terminology isn't obvious unless you're familiar with implementing concurrency primitives (i.e. locks). Normally, it'd be wrapped in a `lock` or `unlock` function, but we can't have real locks since we don't have real threads--we just have interrupt and normal/"process" context. (Real locking either blocks or sleeps under contention, and our interrupt handlers can't do either of those.)

So what's going on here is simply an "unlock" operation, which in itself is trivial, because it's just clearing the flag. However, the barrier is needed to prevent moving the flag change before something else (after inlining the `scpi_free_buffer` function). If you're curious, this has "release" ordering, in the concurrency lingo.

GCC does provide some more-precise helper builtins for the e.g. release semantics (see https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html), but we can't use them because we don't have atomic types. Plus, their names are arguably more confusing--we'd need to use `__atomic_signal_fence(__ATOMIC_RELEASE)`.

Reply:
TL;DR there's no good way to abstract out the concurrency part into a descriptively-named function, and the barrier is named for exactly what it is (and it's the same name used in Linux and elsewhere).

Reply:
Thanks for the detailed explanation.

> it's the same name used in Linux and elsewhere

In that case, it probably makes the most sense to leave it as is.

Reply:
I could improve the comment by switching it around, something like
```
/* Prevent the compiler from moving the store before earlier operations. */
```

Reply:
I think the first comment provided a relatively similar level of description. If you think this one is more descriptive, by all means, please make the change.